batch_builder.go

```diff
@@ -0,0 +1,228 @@
+package hippo
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"sort"
+)
+
+// BatchBuilder is responsible for building a batch that will serialize within a desired size if possible.
+// It'll fit big and small populators together, but doesn't try to do so optimally.
+type BatchBuilder struct {
+	desiredSize       int
+	buf               *bytes.Buffer
+	serializedUpserts [][]byte
+	batchGUID         string
+	replaceAll        bool
+	ttlMinutes        uint32
+	builtBatchesCount int
+	hasClosedBatch    bool
+	sender            TagBatchPartSender // optional - to help track batch origin
+}
+
+// NewBatchBuilder builds a new BatchBuilder
+func NewBatchBuilder(desiredSize int, replaceAll bool, ttlMinutes uint32) *BatchBuilder {
+	return &BatchBuilder{
+		desiredSize:       desiredSize,
+		serializedUpserts: make([][]byte, 0),
+		replaceAll:        replaceAll,
+		ttlMinutes:        ttlMinutes,
+		buf:               bytes.NewBuffer(make([]byte, 0, desiredSize)),
+	}
+}
+
+// Reset resets the batch builder to build a new batch, reusing the underlying buffer.
+func (b *BatchBuilder) Reset(desiredSize int, replaceAll bool, ttlMinutes uint32) {
+	b.desiredSize = desiredSize
+	b.serializedUpserts = b.serializedUpserts[:0]
+	b.batchGUID = ""
+	b.replaceAll = replaceAll
+	b.ttlMinutes = ttlMinutes
+	b.builtBatchesCount = 0
+	b.hasClosedBatch = false
+}
+
+// SetSenderInfo sets optional metadata about the service sending batches
+func (b *BatchBuilder) SetSenderInfo(sender TagBatchPartSender) {
+	b.sender = sender
+}
+
+func (b *BatchBuilder) isSenderInfoSet() bool {
+	return b.sender.ServiceName != "" || b.sender.ServiceInstance != "" || b.sender.HostName != ""
+}
+
+// AddUpsert attempts to add the input upsert into the batch.
+// Make sure to add all upserts to the batch before calling BuildBatch()
+func (b *BatchBuilder) AddUpsert(upsert *TagUpsert) error {
+	ser, err := json.Marshal(upsert)
+	if err != nil {
+		return fmt.Errorf("Error serializing TagUpsert: %s", err)
+	}
+	b.serializedUpserts = append(b.serializedUpserts, ser)
+	return nil
+}
+
+// SetBatchGUID sets the GUID that was returned after submitting the first part to the server.
+func (b *BatchBuilder) SetBatchGUID(guid string) {
+	b.batchGUID = guid
+}
+
+// BuildBatch builds and returns a serialized batch.
+// Once this method is called, don't make any further calls to AddUpsert.
+// - returns serialized batch and upsert count
+func (b *BatchBuilder) BuildBatch() ([]byte, int, error) {
+	if len(b.serializedUpserts) == 0 && b.hasClosedBatch {
+		// nothing to do
+		return nil, 0, nil
+	}
+
+	if b.batchGUID == "" && b.builtBatchesCount > 0 {
+		return nil, 0, fmt.Errorf("Only first batch may be sent without batch GUID")
+	}
+
+	b.buf.Reset()
+
+	// sort upserts by serialized size
+	sort.SliceStable(b.serializedUpserts, func(i int, j int) bool {
+		return len(b.serializedUpserts[i]) < len(b.serializedUpserts[j])
+	})
```

Comment:
This likely doesn't matter too much, but this will be called for every iteration of `BuildBatch` when it really only needs to be called in the first iteration. Once that first iteration has completed we'll have a sorted slice that will just continue to shrink as long as everything goes according to plan. I only point this out given this current approach/implementation seems somewhat fairly focused on performance (based on re-use of an underlying buffer, setting values of slice to nil, shrinking the slice, etc.).

Reply:
Good call - lemme think about it. It's not actually a big deal compared to how much work it's doing, but would be cleaner to update. I'll take a look, thanks

Reply:
Done:
https://github.com/kentik/gohippo/blob/38ed3bf4a4a42d4ce9a67b484791b8f51a1a45d8/batch_builder.go#L90-L95